main.cc

```diff
@@ -0,0 +1,210 @@
+#include <iostream>
+#include <string.h>
+#include <cassert>
+
+using namespace std;
+
+class String {
+public:
+    
+    String() :
+        String("", 0)
+    { }
+
+    String(const char* data) :
+        String(data, strlen(data))
+    { }
+
+    String(const String& s) :
+        String(s.data(), s.length())
+    { }
+    
+    String(const char* data, size_t length) {
+        alloc_and_copy(data, length);
+    }
+    
+    String(unsigned capacity) {
+        m_data = new char[capacity];
+        m_capacity = capacity;
+        m_length = 1;
+    }
+    
+    String(String&& s) :
+        m_data(s.m_data),
+        m_capacity(s.capacity()),
+        m_length(s.length()) 
+    { 
+        s.clean();
+    }
+         
+    String& operator=(String s) {
+        swap(*this, s);
+        return *this;
+    }
+    
+    String& operator=(String&& s) {
+        // for x = x, ignoring more complicated cases where len/capacity differ
+        if (m_data == s.data()) {
+            return *this;
+        }
+        m_data = s.m_data;
+        m_capacity = s.capacity();
+        m_length = s.length();
+        s.clean();
+    }
+        
+    ~String() {
+        delete[] m_data;
+    }
+    
+    void reserve(unsigned capacity) {
+        if (capacity < m_capacity) { return; }
+        reallocate(capacity);
+    }
+    
+    unsigned int length() const {
+        return m_length;
+    }
+
+    unsigned int capacity() const {
+        return m_capacity;
+    }
+
+    const char* data() const {
```

Comment:
The assignment doesn't say that expicitly, but real std::string::data() allows data mutation.